### PR TITLE
Support for Arduino platforms

### PR DESCRIPTION
**Pull request by arengarajan99**
This program adds configurations for 32 bit and Microsecond Support for Arduino Only (can be extended universally by defining target configuration).  Requires a little under 60kB worth of flash storage at the moment (Will not work for Nano and Uno boards as presently constructed). Blink program works out of the box. 

```➜  lingua-franca git:(arduino-support) ✗ bin/lfc test/C/src/arduino/Blink.lf                                              
Generating code for: file:/Users/arengarajan/Documents/GitHub/LinguaFranca/lingua-franca/test/C/src/arduino/Blink.lf
******** mode: STANDALONE
******** generated sources: /Users/arengarajan/Documents/GitHub/LinguaFranca/lingua-franca/test/C/src-gen/arduino/Blink
******** Using 1 threads to compile the program.
Code generation finished.
➜  lingua-franca git:(arduino-support) ✗ arduino-cli compile --fqbn arduino:sam:arduino_due_x test/C/src-gen/arduino/Blink
Sketch uses 54364 bytes (10%) of program storage space. Maximum is 524288 bytes.


Used platform Version Path                                                                     
arduino:sam   1.6.12  /Users/arengarajan/Library/Arduino15/packages/arduino/hardware/sam/1.6.12

➜  lingua-franca git:(arduino-support) ✗ arduino-cli upload --fqbn arduino:sam:arduino_due_x -p /dev/cu.usbmodem1401 test/C/src-gen/arduino/Blink
Atmel SMART device 0x285e0a60 found
Erase flash
done in 0.033 seconds

Write 56796 bytes to flash (222 pages)
[==============================] 100% (222/222 pages)
done in 10.935 seconds
Set boot flash true
CPU reset.
```

### CODE REVIEW

File: test/C/src/arduino/Blink.lf
Comment (petervdonovan) [inline test/C/src/arduino/Blink.lf:21]: Is this comment up to date?

(Also the terminating newline is missing.)

Comment (arengarajan99) [reply test/C/src/arduino/Blink.lf:4]: Sorry, forgot to push the change that updated the comment line. Should be fixed now.

File: org.lflang/src/org/lflang/TargetConfig.java
Comment (arengarajan99) [inline org.lflang/src/org/lflang/TargetConfig.java:264]: Note: These don't have any purpose as of yet as we would need to add configurations for each of these values. Can be apart of a larger commit bringing 32 Bit/microsecond granularity support to Reactor-C

**Comment by lhstrh** [comment]
Did you mean to include changes to the `reactor-cpp` submodule to be part of this PR?

File: org.lflang/src/org/lflang/TargetConfig.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/TargetConfig.java:263]: If it's unused, then let's delete it. Otherwise, this needs Javadoc.

File: org.lflang/src/org/lflang/generator/c/CCompiler.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CCompiler.java:240]: With our Cmake approach, this would just become the same case as `CppMode`, right?

File: org.lflang/src/org/lflang/generator/c/CCoreFilesUtils.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CCoreFilesUtils.java:44]: I believe all our files are in lower case. Let's try to be consistent and make this lowercase as well.

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:1092]: I think its preferred to use `resolve` instead of string concatenations.

File: org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java:80]: What does this comment refer to? If it is something you're not planning to address in this PR, please create an issue and reference it in the `TODO` comment.

**Review by lhstrh — CHANGES_REQUESTED** [review]
Looks good! Provided some feedback.

**Review by cmnrd — DISMISSED** [review]
Looks like this changes the refs of the reactor-cpp and reactor-ts submodules, which shouldn't be affected. Also the ref of reactor-c changes, but this might be expected.

File: org.lflang/src/org/lflang/generator/c/CCompiler.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CCompiler.java]: Yes! Converted over to CPPMode with a few tweaks.

File: org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java]: Changed.

File: .github/workflows/ci.yml
Comment (arengarajan99) [inline .github/workflows/ci.yml:57]: What is the process of CI testing this before we merge the branches together? We can't run our custom c-tests.yml file without specifying the branch but we need to merge these together.

File: .gitmodules
Comment (Soroosh129) [inline .gitmodules:15]: This repository should probably be moved to the lf-lang org.

File: org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java:117]: I don't think copying files should happen here. It should probably happen around [here](https://github.com/lf-lang/lingua-franca/blob/8b35b5937c1529fa9ac36bb34ae1d9d7497c74eb/org.lflang/src/org/lflang/generator/c/CGenerator.java#L567).

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:1091]: Does the Arduino support work with `cmake: false` (i.e., without CMake)? If not, there is no need for this.

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:2256]: We have an `errorReporter` that you could use to report warnings (although I think this should be an error instead of overriding the `threading` target property).

File: org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java
Comment (Soroosh129) [inline org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java:82]: I think all the Arduino-related code can be consolidated into a new class (e.g., called `org.lflang.generator.c.CArduinoSupport.java`) with various static methods (like `generateDefineStatements`). This would make it easy to expand the support infrastructure in the future. However, I don't think this is essential for merging this in.

**Review by Soroosh129 — CHANGES_REQUESTED** [review]
Looks like this is mostly done. I only left a few comments.

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CGenerator.java:2256]: The idea here was to not have to write threading=false inside the lf file since no arduino program will ever have threading.

File: org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CCmakeCompiler.java:117]: Fixed.

File: .gitmodules
Comment (arengarajan99) [reply .gitmodules:15]: Marten requested to keep it as a submodule under the new platform lib directory in case we add more arduino libraries down the line.

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CGenerator.java:2256]: If the behavior was the opposite (threading is false by default, true if enabled), then I think it would make more sense to error.

File: org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java:82]: Can add this down the line when we have more features.

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (arengarajan99) [reply org.lflang/src/org/lflang/generator/c/CGenerator.java:1091]: Taken out since we only support CMake for the time being

File: .gitmodules
Comment (Soroosh129) [reply .gitmodules:15]: Right. My point was that we should move this submodule repo into the lf-lang org (like reactor-c).

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (Soroosh129) [reply org.lflang/src/org/lflang/generator/c/CGenerator.java:2256]: There is a `targetConfig.setByUser` that can help here. If threading is set by user, then I think an error is more appropriate.

File: .gitmodules
Comment (arengarajan99) [reply .gitmodules:15]: Would a fork of the original repository suffice? It hasn't been updated March 2020.

Comment (arengarajan99) [reply .gitmodules:15]: Resolved.

**Review by Soroosh129 — APPROVED** [review]
LGTM!

However, the corresponding PR on reactor-c needs some work it looks like. I think that PR should be merged before this can be merged.

File: .gitmodules
Comment (lhstrh) [reply .gitmodules:15]: @Soroosh129 why do we need to create a fork and reference that? We are not the maintainers of that submodule.

Comment (Soroosh129) [reply .gitmodules:15]: Right, I didn't know that when I made my comment. @arengarajan99 clarified that for me.

File: org.lflang.tests/src/org/lflang/tests/Configurators.java
Comment (lhstrh) [inline org.lflang.tests/src/org/lflang/tests/Configurators.java:87]: This doesn't look right. Arduino doesn't belong in the list of categories that are not compatible with single-threaded execution.

File: org.lflang.tests/src/org/lflang/tests/runtime/CCppTest.java
Comment (lhstrh) [inline org.lflang.tests/src/org/lflang/tests/runtime/CCppTest.java:46]: I assume Arduino excluded from `CCpp` because we don't want the tests twice?

File: org.lflang/src/org/lflang/generator/c/CCmakeGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CCmakeGenerator.java:245]:
```suggestion
        if (this.targetConfig.platform == Platform.ARDUINO) {
```

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:559]:
```suggestion
                if (targetConfig.platform == Platform.ARDUINO) {
```

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:2256]:
```suggestion
        if (targetConfig.threading && targetConfig.platform == Platform.ARDUINO) {
```

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:2259]:
```suggestion
            if (targetConfig.setByUser.contains(TargetProperty.THREADING)) {
```

File: org.lflang/src/org/lflang/generator/c/CMainGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CMainGenerator.java:49]: Has this FIXME been addressed yet?

File: org.lflang/src/org/lflang/generator/c/CMainGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CMainGenerator.java:44]: Outdated?

File: test/C/src/arduino/AnalogReadSerial.lf
Comment (lhstrh) [inline test/C/src/arduino/AnalogReadSerial.lf:14]:
```suggestion
 }
 
```

File: test/C/src/arduino/BlinkAttemptThreading.lf
Comment (lhstrh) [inline test/C/src/arduino/BlinkAttemptThreading.lf:23]:
```suggestion
 }
 
```

File: test/C/src/arduino/BoardOptions.cmake
Comment (lhstrh) [inline test/C/src/arduino/BoardOptions.cmake:14]: I suppose this is meant to be the default?

File: test/C/src/arduino/BoardOptions.cmake
Comment (lhstrh) [inline test/C/src/arduino/BoardOptions.cmake:63]: It looks like multiple processor options are uncommented. Why?

File: test/C/src/arduino/DigitalReadSerial.lf
Comment (lhstrh) [inline test/C/src/arduino/DigitalReadSerial.lf:17]:
```suggestion
 }
 
```

File: test/C/src/arduino/Fade.lf
Comment (lhstrh) [inline test/C/src/arduino/Fade.lf:26]:
```suggestion
 }
 
```

File: test/C/src/arduino/ReadAnalogVoltage.lf
Comment (lhstrh) [inline test/C/src/arduino/ReadAnalogVoltage.lf:21]:
```suggestion
 }
 
```

**Review by lhstrh — CHANGES_REQUESTED** [review]
Looks like this needs a bit more fine tuning. Also, you've accidentally updated the `reactor-ts` submodule as part of this PR.

File: org.lflang/src/org/lflang/generator/c/CGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CGenerator.java:560]: Exceeding max line width.

File: org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CPreambleGenerator.java:81]:
```suggestion
        if (targetConfig.platform == Platform.ARDUINO) {
```

**Review by lhstrh — APPROVED** [review]
Forgot to hit the `Finish your review` button a while ago. Apologies! Here is some dangling feedback.

File: org.lflang/src/org/lflang/generator/c/CMainGenerator.java
Comment (lhstrh) [reply org.lflang/src/org/lflang/generator/c/CMainGenerator.java:44]: Action item: clarify comment.

**Pull request merged**
